libcaf_net/caf/net/endpoint_manager_impl.hpp

```diff
@@ -87,25 +89,36 @@ class endpoint_manager_impl : public endpoint_manager {
   }
 
   bool handle_write_event() override {
-    if (!this->events_.blocked() && !this->events_.empty()) {
+    if (!this->queue_.blocked()) {
+      this->queue_.fetch_more();
+      auto& q = std::get<0>(this->queue_.queue().queues());
       do {
-        this->events_.fetch_more();
-        auto& q = this->events_.queue();
         q.inc_deficit(q.total_task_size());
         for (auto ptr = q.next(); ptr != nullptr; ptr = q.next()) {
-          using timeout = endpoint_manager::event::timeout;
-          using resolve_request = endpoint_manager::event::resolve_request;
-          if (auto rr = get_if<resolve_request>(&ptr->value)) {
-            transport_.resolve(*this, std::move(rr->path),
-                               std::move(rr->listener));
-          } else {
-            auto& t = get<timeout>(ptr->value);
-            transport_.timeout(*this, t.type, t.id);
-          }
+          auto f = detail::make_overload(
+            [&](endpoint_manager_queue::event::resolve_request& x) {
+              transport_.resolve(*this, x.locator, x.listener);
+            },
+            [&](endpoint_manager_queue::event::new_proxy& x) {
+              transport_.new_proxy(*this, x.peer, x.id);
+            },
+            [&](endpoint_manager_queue::event::local_actor_down& x) {
+              transport_.local_actor_down(*this, x.observing_peer, x.id,
+                                          std::move(x.reason));
+            },
+            [&](endpoint_manager_queue::event::timeout& x) {
+              transport_.timeout(*this, x.type, x.id);
+            });
+          visit(f, ptr->value);
         }
-      } while (!this->events_.try_block());
+      } while (!q.empty());
     }
-    return transport_.handle_write_event(*this);
+    if (!transport_.handle_write_event(*this)) {
+      if (this->queue_.blocked())
+        return false;
```

Comment:
I'm simply not entirely sure what's happening here. Would you mind summarizing it in a few words. I mean, the manager somehow dequeues events and handles them before passing the event to its transport, but I'm a bit unsure what the deficit + for loop does. Does this query the event queue before letting the transport handle the message queue or ...?

Reply:
The intrusive CAF queues implement [deficit round robin](https://en.wikipedia.org/wiki/Deficit_round_robin). By setting `q.inc_deficit(q.total_task_size());` before entering the loop we make sure that we can consume the entire thing at once.

That's of course hacky, but it's not meant as the strategy we finally want to have. Eventually, we can make use of DRR to only fetch messages as long as we have some space on the socket (maybe plus some extra buffering) for writing.